### DISCUSSION TITLE
OTEL Automatic Tracing for Micronaut 4 + Hibernate Reactive + Vert.x

### DISCUSSION THREAD

**Discussion by s-af** [micronaut-projects/micronaut-data#2516] [General]
Hi,

In my current setup I am using Micronaut 4 + Hibernate Reactive. I was trying to enable OTEL tracing for DB operations using java Agent (https://github.com/open-telemetry/opentelemetry-java-instrumentation). Starting with version 1.30.0 OTEL Java Agent instrument Hibernate Reactive and Vert.x operations. Unfortunately I was not able make this combination work. Hibernate traces not appears at all. Vert.x traces appears without link to parent.
After some debugging I found that Micronaut use SessionFactory.openSession() to obtains session. From other side OTEL Hibernate Reactive instrumentation instrument SessionFactory.withSession(), but doesn't instrument SessionFactory.openSession(). Based on Hibernate documentation both of this cases definitely valid, but withSession() is preferred way because in that case Hibernate will take care about Vert.x local context. 
That is what Hibernate documentation say:
_"When you create a session using withSession() or withTransaction(), it’s automatically associated with the current Vert.x local context and propagates with the local context."_
_"On the other hand, if you use openSession(), you’ll have to manage the association between sessions and contexts yourself."_

So my questions are:
- Is it possible somehow to propagate reactor context to Vert.x context (e.g. trace-id, span-id) using Micronaut API? It probably should fix Vert.x tracing.
- Is it possible to use withSession() instead of openSession() to obtain sessions in Micronaut Data? It probably should fix Hibernate Reactive & Vert.x tracing.

Any ideas guys?

**Comment by dstepanov** [comment]
First of all, it's not that easy to do propagation and tracing in the reactive/async programming, and honestly I'm not sure if the instrumentation can properly support it.

1. I don't see how it would be possible
2. We don't use those methods, because we do propagate the session in our own context

The only solution would be for us to support tracing Micronaut Data query executions and include them in the OpenTelemetry context.

**Reply by s-af** [reply]
Thank you for response. 
Issue fixed in hibernate-reactive instrumentation 
https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/9474#discussioncomment-7031663
